File: r/sedonadb/R/context.R
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

#' Create a SedonaDB context
#'
#' Runtime options configure the execution environment. Use
#' `global = TRUE` to configure the global context or use the
#' returned object as a scoped context. A scoped context is
#' recommended for programmatic usage as it prevents named
#' views from interfering with each other.
#'
#' @param global Use TRUE to set options on the global context.
#' @param memory_limit Maximum memory for query execution, as a
#'   human-readable string (e.g., `"4gb"`, `"512m"`) or `NULL` for
#'   unbounded (the default).
#' @param temp_dir Directory for temporary/spill files, or `NULL` to
#'   use the DataFusion default.
#' @param memory_pool_type Memory pool type: `"greedy"` (default) or
#'   `"fair"`. Only takes effect when `memory_limit` is set.
#' @param unspillable_reserve_ratio Fraction of memory (0--1) reserved for
#'   unspillable consumers. Only applies when `memory_pool_type` is
#'   `"fair"`. Defaults to 0.2 when not explicitly set.
#' @param ... Reserved for future options
#'
#' @returns The constructed context, invisibly.
#' @export
#'
#' @examples
#' sd_connect(memory_limit = "100mb", memory_pool_type = "fair")
#'
#'
sd_connect <- function(
  ...,
  global = FALSE,
  memory_limit = NULL,
  temp_dir = NULL,
  memory_pool_type = NULL,
  unspillable_reserve_ratio = NULL
) {
  check_dots_empty(..., label = "sd_connect()")
  options <- list()

  if (!is.null(memory_limit)) {
    stopifnot(is.character(memory_limit), length(memory_limit) == 1L)
    options[["memory_limit"]] <- memory_limit
  }

  if (!is.null(temp_dir)) {
    stopifnot(is.character(temp_dir), length(temp_dir) == 1L)
    options[["temp_dir"]] <- temp_dir
  }

  if (!is.null(memory_pool_type)) {
    memory_pool_type <- match.arg(memory_pool_type, c("greedy", "fair"))
    options[["memory_pool_type"]] <- memory_pool_type
  }

  if (!is.null(unspillable_reserve_ratio)) {
    stopifnot(
      is.numeric(unspillable_reserve_ratio),
      length(unspillable_reserve_ratio) == 1L,
      unspillable_reserve_ratio >= 0,
      unspillable_reserve_ratio <= 1
    )
    options[["unspillable_reserve_ratio"]] <- as.character(
      unspillable_reserve_ratio
    )
  }

  # Don't replace the global context
  if (global && length(options) > 0 && !is.null(global_ctx$ctx)) {
    warning(
      "Cannot change runtime options after the context has been initialized. ",
      "Set global options with sd_connect() before executing your first query ",
      "or use a scoped context by passing global = FALSE"
    )

    return(invisible(global_ctx$ctx))
  } else if (global && !is.null(global_ctx$ctx)) {
    return(invisible(global_ctx$ctx))
  }

  keys <- as.character(names(options))
  values <- as.character(options)
  ctx <- InternalContext$new(keys, values)

  if (global) {
    global_ctx$ctx <- ctx
  }

  invisible(ctx)
}

#' Create a DataFrame from one or more Parquet files
#'
#' The query will only be executed when requested.
#'
#' @param ctx A SedonaDB context.
#' @param path One or more paths or URIs to Parquet files
#' @param options A named list of options to pass to the Parquet reader.
#' @param geometry_columns A JSON string or named list mapping column names to
#'   GeoParquet column metadata. This can be used to mark binary WKB columns as
#'   geometry columns or override existing GeoParquet metadata.
#' @param validate Whether to validate geometry column contents against their
#'   metadata. Currently, this validates WKB input.
#' @param partitioning A character vector of column names for hive-style
#'   partitioning. By default (`NULL`), partition columns are auto-discovered.
#'   Use `character()` to disable partitioning.
#'
#' @returns A sedonadb_dataframe
#' @export
#'
#' @examples
#' path <- system.file("files/natural-earth_cities_geo.parquet", package = "sedonadb")
#' sd_read_parquet(path) |> head(5) |> sd_preview()
#'
sd_read_parquet <- function(
  path,
  options = NULL,
  geometry_columns = NULL,
  validate = FALSE,
  partitioning = NULL
) {
  sd_ctx_read_parquet(
    ctx(),
    path,
    options = options,
    geometry_columns = geometry_columns,
    validate = validate,
    partitioning = partitioning
  )
}

#' @rdname sd_read_parquet
#' @export
sd_ctx_read_parquet <- function(
  ctx,
  path,
  options = NULL,
  geometry_columns = NULL,
  validate = FALSE,
  partitioning = NULL
) {
  check_ctx(ctx)

  if (is.null(options)) {
    options <- list()
  } else {
    options <- Filter(Negate(is.null), as.list(options))
  }

  option_keys <- names(options)
  if (
    length(options) > 0L &&
      (is.null(option_keys) || any(is.na(option_keys)) || any(option_keys == ""))
  ) {
    stop("All option values must be named")
  }

  if (length(options) == 0L) {
    option_keys <- character()
  }

  option_values <- vapply(
    options,
    function(value) {
      if (length(value) != 1L) {
        stop("All option values must be length 1")
      }

      as.character(value)
    },
    character(1)
  )

  if (is.list(geometry_columns)) {
    geometry_columns <- as.character(jsonlite::toJSON(
      geometry_columns,
      auto_unbox = TRUE,
      null = "null"
    ))
  } else if (!is.null(geometry_columns)) {
    if (
      !is.character(geometry_columns) ||
        length(geometry_columns) != 1L ||
        is.na(geometry_columns)
    ) {
      stop("`geometry_columns` must be a JSON string, named list, or NULL")
    }
  }

  if (!is.logical(validate) || length(validate) != 1L || is.na(validate)) {
    stop("`validate` must be TRUE or FALSE")
  }

  if (
    !is.null(partitioning) &&
      (!is.character(partitioning) || any(is.na(partitioning)))
  ) {
    stop("`partitioning` must be a character vector or NULL")
  }

  df <- ctx$read_parquet(
    path,
    option_keys,
    unname(option_values),
    geometry_columns,
    validate,
    partitioning
  )
  new_sedonadb_dataframe(ctx, df)
}

#' Create a DataFrame from SQL
#'
#' The query will only be executed when requested.
#'
#' @param ctx A SedonaDB context.
#' @param sql A SQL string to execute
#' @param params A list of parameters to fill placeholders in the query.
#' @param ... These dots are for future extensions and currently must be empty.
#'
#' @returns A sedonadb_dataframe
#' @export
#'
#' @examples
#' sd_sql("SELECT ST_Point(0, 1) as geom")
#' sd_sql("SELECT ST_Point($1, $2) as geom", params = list(1, 2))
#' sd_sql("SELECT ST_Point($x, $y) as geom", params = list(x = 1, y = 2))
#'
sd_sql <- function(sql, ..., params = NULL) {
  sd_ctx_sql(ctx(), sql, ..., params = params)
}

#' @rdname sd_sql
#' @export
sd_ctx_sql <- function(ctx, sql, ..., params = NULL) {
  check_ctx(ctx)
  df <- ctx$sql(sql)
  check_dots_empty(..., label = "sd_sql()")

  if (!is.null(params)) {
    params <- lapply(params, as_sedonadb_literal)
    df <- df$with_params(params)
  }

  new_sedonadb_dataframe(ctx, df)
}

#' Create or Drop a named view
#'
#' Remove a view created with [sd_to_view()] from the context.
#'
#' @param ctx A SedonaDB context.
#' @param table_ref The name of the view reference
#' @returns The context, invisibly
#' @export
#'
#' @examples
#' sd_sql("SELECT 1 as one") |> sd_to_view("foofy")
#' sd_view("foofy")
#' sd_drop_view("foofy")
#' try(sd_view("foofy"))
#'
sd_drop_view <- function(table_ref) {
  sd_ctx_drop_view(ctx(), table_ref)
}

#' @rdname sd_drop_view
#' @export
sd_ctx_drop_view <- function(ctx, table_ref) {
  check_ctx(ctx)
  ctx$deregister_table(table_ref)
  invisible(ctx)
}

#' @rdname sd_drop_view
#' @export
sd_view <- function(table_ref) {
  sd_ctx_view(ctx(), table_ref)
}

#' @rdname sd_drop_view
#' @export
sd_ctx_view <- function(ctx, table_ref) {
  check_ctx(ctx)
  df <- ctx$view(table_ref)
  new_sedonadb_dataframe(ctx, df)
}

# nolint start: line_length_linter
#' Register a user-defined function
#'
#' Several types of user-defined functions can be registered into a session
#' context. Currently, the only implemented variety is an external pointer
#' to a Rust `FFI_ScalarUDF`, an example of which is available from the
#' [DataFusion Python documentation](https://github.com/apache/datafusion-python/blob/6f3b1cab75cfaa0cdf914f9b6fa023cb9afccd7d/examples/datafusion-ffi-example/src/scalar_udf.rs).
#'
#' @param ctx A SedonaDB context.
#' @param udf An object of class 'datafusion_scalar_udf'
#'
#' @returns NULL, invisibly
#' @export
#'
sd_register_udf <- function(udf) {
  sd_ctx_register_udf(ctx(), udf)
}

#' @rdname sd_register_udf
#' @export
sd_ctx_register_udf <- function(ctx, udf) {
  check_ctx(ctx)
  ctx$register_scalar_udf(udf)
}
# nolint end

# We mostly use one context but support isolated contexts via sd_ctx_*
# functions, which more cleanly scope the registration of views and
# UDFs.
ctx <- function() {
  if (is.null(global_ctx$ctx)) {
    global_ctx$ctx <- sd_connect()
  }

  global_ctx$ctx
}

check_ctx <- function(ctx) {
  if (!inherits(ctx, "sedonadb::InternalContext")) {
    stop(
      sprintf(
        "Expected ctx to be a SedonaDB context but got object of class '%s'",
        class(ctx)[1]
      )
    )
  }

  invisible(ctx)
}

global_ctx <- new.env(parent = emptyenv())
global_ctx$ctx <- NULL


#' Configure PROJ
#'
#' Performs a runtime configuration of PROJ, which can be used in place of
#' a build-time linked version of PROJ or to add in support if PROJ was
#' not linked at build time.
#'
#' @param preset One of:
#'   - `"homebrew"`: Look for PROJ installed by Homebrew. This is the easiest
#'     option on MacOS.
#'   - `"system"`: Look for PROJ in the platform library load path (e.g.,
#'     after installing system proj on Linux).
#'   - `"auto"`: Try all presets in the order listed above, issuing a warning
#'     if none can be configured.
#' @param shared_library An absolute or relative path to a shared library
#'   valid for the platform.
#' @param database_path A path to proj.db
#' @param search_path A path to the data files required by PROJ for some
#'   transforms.
#'
#' @returns NULL, invisibly
#' @export
#'
#' @examples
#' sd_configure_proj("auto")
#'
sd_configure_proj <- function(
  preset = NULL,
  shared_library = NULL,
  database_path = NULL,
  search_path = NULL
) {
  if (!is.null(preset)) {
    switch(
      preset,
      homebrew = {
        configure_proj_prefix(Sys.getenv("HOMEBREW_PREFIX", "/opt/homebrew"))
        return(invisible(NULL))
      },
      system = {
        configure_proj_system()
        return(invisible(NULL))
      },
      auto = {
        presets <- c("homebrew", "system")
        errors <- c()
        for (preset in presets) {
          maybe_err <- try(sd_configure_proj(preset), silent = TRUE)
          if (!inherits(maybe_err, "try-error")) {
            return(invisible(NULL))
          } else {
            errors <- c(errors, sprintf("%s: %s", preset, maybe_err))
          }
        }

        packageStartupMessage(
          sprintf(
            "Failed to configure PROJ (tried %s):\n%s",
            paste0("'", presets, "'", collapse = ", "),
            paste0(errors, collapse = "\n")
          )
        )

        return(invisible(NULL))
      },
      stop(sprintf("Unknown preset: '%s'", preset))
    )
  }

  # We could check a shared library with dyn.load(), but this may error for
  # valid system PROJ that isn't an absolute filename.

  if (!is.null(database_path)) {
    if (!file.exists(database_path)) {
      stop(sprintf("Invalid database path: '%s' does not exist", database_path))
    }
  }

  if (!is.null(search_path)) {
    if (!dir.exists(search_path)) {
      stop(sprintf("Invalid search path: '%s' does not exist", search_path))
    }
  }

  configure_proj_shared(
    shared_library_path = shared_library,
    database_path = database_path,
    search_path = search_path
  )
}

configure_proj_system <- function() {
  sd_configure_proj(shared_library = proj_dll_name())
}

configure_proj_prefix <- function(prefix) {
  if (!dir.exists(prefix)) {
    stop(sprintf("Can't configure PROJ from prefix '%s': does not exist", prefix))
  }

  sd_configure_proj(
    shared_library = file.path(prefix, "lib", proj_dll_name()),
    database_path = file.path(prefix, "share", "proj", "proj.db"),
    search_path = file.path(prefix, "share", "proj")
  )
}

proj_dll_name <- function() {
  switch(
    tolower(Sys.info()[["sysname"]]),
    windows = "proj.dll",
    darwin = "libproj.dylib",
    linux = "libproj.so",
    stop(sprintf(
      "Can't determine system PROJ shared library name for OS: %s",
      Sys.info()[["sysname"]]
    ))
  )
}

File: r/sedonadb/src/rust/src/context.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
use std::collections::HashMap;
use std::sync::Arc;

use arrow_array::RecordBatchReader;
use arrow_schema::{ArrowError, DataType};
use datafusion::catalog::{MemTable, TableProvider};
use datafusion_ffi::udf::FFI_ScalarUDF;
use savvy::{savvy, savvy_err, IntoExtPtrSexp, NotAvailableValue, OwnedStringSexp, Result};
use serde_json::{Map, Number, Value};

use sedona::{
    context::SedonaContext, context_builder::SedonaContextBuilder,
    record_batch_reader_provider::RecordBatchReaderProvider,
};
use sedona_extension::runtime::RuntimeHandle;
use sedona_geoparquet::provider::GeoParquetReadOptions;

use crate::{
    dataframe::{new_data_frame, InternalDataFrame},
    ffi::{import_array_stream, import_scalar_udf, import_table_provider, FFIScalarUdfR},
    runtime::wait_for_future_captured_r,
};

fn json_scalar_or_array(values: Vec<Value>) -> Value {
    if values.len() == 1 {
        values.into_iter().next().unwrap()
    } else {
        Value::Array(values)
    }
}

fn r_value_to_json(value: savvy::Sexp) -> Result<Value> {
    if value.is_null() {
        return Ok(Value::Null);
    }

    if value.is_list() {
        let values = savvy::ListSexp::try_from(value)?;
        let names = values.names_iter().collect::<Vec<_>>();
        let has_names = names.iter().any(|name| !name.is_empty());

        if has_names {
            if names.iter().any(|name| name.is_empty()) {
                return Err(savvy_err!(
                    "JSON object values must be either all named or all unnamed"
                ));
            }

            let mut object = Map::new();
            for (name, item) in names.into_iter().zip(values.values_iter()) {
                if object
                    .insert(name.to_string(), r_value_to_json(item)?)
                    .is_some()
                {
                    return Err(savvy_err!("JSON object names must be unique"));
                }
            }
            return Ok(Value::Object(object));
        }

        return values
            .values_iter()
            .map(r_value_to_json)
            .collect::<Result<Vec<_>>>()
            .map(Value::Array);
    }

    if value.is_logical() {
        let values = savvy::LogicalSexp::try_from(value)?
            .as_slice_raw()
            .iter()
            .map(|value| {
                if value.is_na() {
                    Value::Null
                } else {
                    Value::Bool(*value != 0)
                }
            })
            .collect();
        return Ok(json_scalar_or_array(values));
    }

    if value.is_integer() {
        let values = savvy::IntegerSexp::try_from(value)?
            .iter()
            .map(|value| {
                if value.is_na() {
                    Value::Null
                } else {
                    Value::Number(Number::from(*value))
                }
            })
            .collect();
        return Ok(json_scalar_or_array(values));
    }

    if value.is_real() {
        let values = savvy::RealSexp::try_from(value)?
            .iter()
            .map(|value| {
                if value.is_na() {
                    Ok(Value::Null)
                } else {
                    Number::from_f64(*value)
                        .map(Value::Number)
                        .ok_or_else(|| savvy_err!("JSON numbers must be finite"))
                }
            })
            .collect::<Result<Vec<_>>>()?;
        return Ok(json_scalar_or_array(values));
    }

    if value.is_string() {
        let values = savvy::StringSexp::try_from(value)?
            .iter()
            .map(|value| {
                if value.is_na() {
                    Value::Null
                } else {
                    Value::String(value.to_string())
                }
            })
            .collect();
        return Ok(json_scalar_or_array(values));
    }

    Err(savvy_err!(
        "Cannot convert R type '{}' to JSON",
        value.get_human_readable_type_name()
    ))
}

#[savvy]
pub struct InternalContext {
    pub inner: Arc<SedonaContext>,
    pub runtime: Arc<RuntimeHandle>,
}

#[savvy]
impl InternalContext {
    pub fn new(option_keys: savvy::Sexp, option_values: savvy::Sexp) -> Result<Self> {
        let runtime = tokio::runtime::Builder::new_multi_thread()
            .enable_all()
            .build()?;

        let keys = savvy::StringSexp::try_from(option_keys)?;
        let values = savvy::StringSexp::try_from(option_values)?;

        let options: HashMap<String, String> = keys
            .iter()
            .zip(values.iter())
            .map(|(k, v)| (k.to_string(), v.to_string()))
            .collect();

        let inner = if options.is_empty() {
            wait_for_future_captured_r(&runtime, SedonaContext::new_local_interactive())??
        } else {
            let builder = SedonaContextBuilder::from_options(&options)?;
            wait_for_future_captured_r(&runtime, builder.build())??
        };

        Ok(Self {
            inner: Arc::new(inner),
            runtime: Arc::new(RuntimeHandle::new(runtime)),
        })
    }

    #[allow(clippy::too_many_arguments)]
    pub fn read_parquet(
        &self,
        paths: savvy::Sexp,
        option_keys: savvy::Sexp,
        option_values: savvy::Sexp,
        geometry_columns: savvy::Sexp,
        validate: bool,
        partitioning: savvy::Sexp,
    ) -> Result<InternalDataFrame> {
        let paths_strsxp = savvy::StringSexp::try_from(paths)?;
        let table_paths = paths_strsxp
            .iter()
            .map(|s| s.to_string())
            .collect::<Vec<_>>();

        let option_keys_strsxp = savvy::StringSexp::try_from(option_keys)?;
        let option_values_strsxp = savvy::StringSexp::try_from(option_values)?;
        let options = option_keys_strsxp
            .iter()
            .zip(option_values_strsxp.iter())
            .map(|(key, value)| (key.to_string(), value.to_string()))
            .collect::<HashMap<_, _>>();

        let mut geo_options = GeoParquetReadOptions::from_table_options(options)
            .map_err(|e| savvy_err!("Invalid table options: {e}"))?;

        if !geometry_columns.is_null() {
            let geometry_columns_json = if geometry_columns.is_list() {
                serde_json::to_string(&r_value_to_json(geometry_columns)?)
                    .map_err(|e| savvy_err!("Failed to serialize geometry_columns: {e}"))?
            } else {
                let geometry_columns_strsxp = savvy::StringSexp::try_from(geometry_columns)?;
                geometry_columns_strsxp
                    .iter()
                    .next()
                    .ok_or_else(|| savvy_err!("`geometry_columns` must have length 1"))?
                    .to_string()
            };
            geo_options = geo_options
                .with_geometry_columns_json(&geometry_columns_json)
                .map_err(|e| savvy_err!("Invalid geometry_columns JSON: {e}"))?;
        }

        geo_options = geo_options.with_validate(validate);

        if !partitioning.is_null() {
            let partitioning_strsxp = savvy::StringSexp::try_from(partitioning)?;
            geo_options = geo_options.with_table_partition_cols(
                partitioning_strsxp
                    .iter()
                    .map(|name| (name.to_string(), DataType::Utf8View))
                    .collect(),
            );
        }

        let inner_context = self.inner.clone();
        let inner = wait_for_future_captured_r(&self.runtime, async move {
            inner_context.read_parquet(table_paths, geo_options).await
        })??;

        Ok(new_data_frame(inner, self.runtime.clone()))
    }

    pub fn sql(&self, query: &str) -> Result<InternalDataFrame> {
        let query_string = query.to_string();
        let inner_context = self.inner.clone();
        let inner = wait_for_future_captured_r(&self.runtime, async move {
            inner_context.sql(&query_string).await
        })??;
        Ok(new_data_frame(inner, self.runtime.clone()))
    }

    pub fn view(&self, table_ref: &str) -> Result<InternalDataFrame> {
        let inner_context = self.inner.clone();
        let table_ref_string = table_ref.to_string();
        let inner = wait_for_future_captured_r(&self.runtime, async move {
            inner_context.ctx.table(table_ref_string).await
        })??;
        Ok(new_data_frame(inner, self.runtime.clone()))
    }

    pub fn data_frame_from_array_stream(
        &self,
        stream_xptr: savvy::Sexp,
        collect_now: bool,
    ) -> savvy::Result<InternalDataFrame> {
        let stream_reader = import_array_stream(stream_xptr)?;

        // Some readers are sensitive to being collected on the R thread or not, so
        // provide the option to collect everything immediately.
        let provider: Arc<dyn TableProvider> = if collect_now {
            let schema = stream_reader.schema();
            let batches = stream_reader.collect::<std::result::Result<Vec<_>, ArrowError>>()?;
            Arc::new(MemTable::try_new(schema, vec![batches])?)
        } else {
            Arc::new(RecordBatchReaderProvider::new(Box::new(stream_reader)))
        };

        let inner = self.inner.ctx.read_table(provider)?;
        Ok(new_data_frame(inner, self.runtime.clone()))
    }

    pub fn data_frame_from_table_provider(
        &self,
        provider_xptr: savvy::Sexp,
    ) -> Result<InternalDataFrame> {
        let provider = import_table_provider(provider_xptr)?;
        let inner = self.inner.ctx.read_table(provider)?;
        Ok(new_data_frame(inner, self.runtime.clone()))
    }

    pub fn deregister_table(&self, table_ref: &str) -> savvy::Result<()> {
        self.inner.ctx.deregister_table(table_ref)?;
        Ok(())
    }

    pub fn list_functions(&self) -> savvy::Result<savvy::Sexp> {
        let mut fn_names = Vec::new();
        let state = self.inner.ctx.state();
        fn_names.extend(state.scalar_functions().keys());
        fn_names.extend(state.aggregate_functions().keys());

        let fn_names_sexp = OwnedStringSexp::try_from(fn_names.as_slice())?;
        fn_names_sexp.into()
    }

    pub fn scalar_udf_xptr(&self, name: &str) -> savvy::Result<savvy::Sexp> {
        if let Some(udf) = self.inner.ctx.state().scalar_functions().get(name) {
            let ffi_scalar_udf: FFI_ScalarUDF = udf.clone().into();
            let mut ffi_xptr = FFIScalarUdfR(ffi_scalar_udf).into_external_pointer();
            unsafe { savvy_ffi::Rf_protect(ffi_xptr.0) };
            ffi_xptr.set_class(vec!["datafusion_scalar_udf"])?;
            unsafe { savvy_ffi::Rf_unprotect(1) };

            Ok(ffi_xptr)
        } else {
            Err(savvy_err!("Scalar UDF '{name}' was not found"))
        }
    }

    pub fn register_scalar_udf(&self, scalar_udf_xptr: savvy::Sexp) -> savvy::Result<()> {
        let scalar_udf = import_scalar_udf(scalar_udf_xptr)?;
        self.inner.ctx.register_udf(scalar_udf);
        Ok(())
    }
}
